stdlib/src/builtin/sort.mojo

```diff
@@ -15,383 +15,175 @@
-        var value = array[i]
-        var j = i
-
-        # Find the placement of the value in the array, shifting as we try to
-        # find the position. Throughout, we assume array[start:i] has already
-        # been sorted.
-        while j > start and not cmp_fn[type](array[j - 1], value):
-            array[j] = array[j - 1]
-            j -= 1
-
-        array[j] = value
+# ===----------------------------------------------------------------------=== #
+#  Scalar list sorting
+# ===----------------------------------------------------------------------=== #
 
 
-fn _insertion_sort[
-    type: CollectionElement, cmp_fn: fn (type, type) capturing -> Bool
-](array: UnsafePointer[type], start: Int, end: Int):
-    """Sort the array[start:end] slice"""
+@always_inline
+fn insertion_sort[dtype: DType](inout list: List[Scalar[dtype]]):
+    """Sort list of scalars in-place with insertion sort algorithm.
 
-    for i in range(start + 1, end):
-        var value = array[i]
-        var j = i
+    Parameters:
+        dtype: The dtype of the scalar.
 
-        # Find the placement of the value in the array, shifting as we try to
-        # find the position. Throughout, we assume array[start:i] has already
-        # been sorted.
-        while j > start and not cmp_fn(array[j - 1], value):
-            array[j] = array[j - 1]
+    Args:
+        list: The list of the scalars which will be sorted in-place.
+    """
+    for i in range(1, len(list)):
+        var key = list[i]
+        var j = i - 1
+        while j >= 0 and key < list[j]:
+            list[j + 1] = list[j]
             j -= 1
+        list[j + 1] = key
 
-        array[j] = value
-
-
-@always_inline
-fn _partition[
-    type: AnyRegType, cmp_fn: _cmp_fn_type
-](array: Pointer[type], start: Int, end: Int) -> Int:
-    if start == end:
-        return end
 
-    var pivot = start + (end - start) // 2
+fn _quick_sort[
+    dtype: DType
+](inout list: List[Scalar[dtype]], low: Int, high: Int):
+    """Sort section of the list, between low and high, with quick sort algorithm in-place.
 
-    var pivot_value = array[pivot]
-
-    var left = start
-    var right = end - 2
-
-    swap(array[pivot], array[end - 1])
+    Parameters:
+        dtype: The dtype of the scalar.
 
-    while left < right:
-        if cmp_fn[type](array[left], pivot_value):
-            left += 1
-        elif not cmp_fn[type](array[right], pivot_value):
-            right -= 1
-        else:
-            swap(array[left], array[right])
+    Args:
+        list: The list of the scalars which will be sorted in-place.
+        low: Int value identifying the lowest index of the list section to be sorted.
+        high: Int value identifying the highest index of the list section to be sorted.
+    """
 
-    if cmp_fn[type](array[right], pivot_value):
-        right += 1
-    swap(array[end - 1], array[right])
-    return right
+    @always_inline
+    @parameter
+    fn _partition(low: Int, high: Int) -> Int:
```

Comment:
**Suggestion** Do you mind lifting this out of `_quick_sort` as a free function?  We use `_partition` internally  and it would be nice to preserve that level of flexibility for custom sorting things.

Reply:
@JoeLoser I just wanted to lift the `_partition` function to upper level, but then I looked through the previous code again and I noticed that the top level `partition` and `_partition` functions  expect `cmp_fn: _cmp_fn_type` as a parameter, which I think is quite specific for the `builtin` sort module. Or rather, we need to decide if we want a complete set of functions which expect the `cmp_fn: _cmp_fn_type` parameter. Currently I implemented a set of functions for `Scalar` list elements (mainly because allows some performance optimisations) and  `ComparableCollectionElement` list elements. 

If we want to keep 

`alias _cmp_fn_type = fn[type: AnyRegType] (type, type) capturing -> Bool`

It will make sense to introduce a third set of functions just with this signature. To be honest I would be agains it as it increases maintenance complexity for a very specific use case, but I am ok with going along with it, if it important internally.

Reply:
Similar to my other comments, I think we should introduce a generic range partition that works on `ComparableCollectionElement` I suppose.  I'm imagining something similar to `std::ranges::partition` from C++ [doc](https://en.cppreference.com/w/cpp/algorithm/ranges/partition).  What do you think about that?  I think partition as a general algorithm (outside of just the sorting use case) is useful and it'll evolve over time.